[Tang Swen Yi] iP

EssWhyy opened the pull request:
###  Mr. Meeseeks Task Bot
```
                                        _       
  _ __  _ _   _ __  ___ ___ ___ ___ ___| |__ ___
 | '  \| '_| | '  \/ -_) -_|_-</ -_) -_) / /(_-<
 |_|_|_|_|   |_|_|_\___\___/__/\___\___|_\_\/__/
```

Introducing the Mr. Meeseeks Task Bot! Based of its original counterpart from [Rick & Morty](https://rickandmorty.fandom.com/wiki/Mr._Meeseeks), the Mr. Meeseeks Bot helps to:

- Keep track and manipulate any and every task you feed it
- Then *poofs* out of existence when it's done!

> _"I'm Mr. Meeseeks look at me!"_ - Mr. Meeseeks Bot

**Some features of Mr. Meeseeks Task Bot include:**

- [X] Add, delete and mark 3 different types of tasks 
- [X] Save tasks in a local file for later to be loaded
- [X] Search for tasks by keywords
- [ ] Cool Graphical User Interface with avatars (Coming Soon)

**How do you use Mr. Meeseeks Bot?**

1. Download the repo onto your local machine
2. Open the Duke.java file in terminal/Command line (cd to correct path, then `java Duke.java`)
3. Mr Meeseeks will help you with the rest! 😎

Comment on this peer review if you have any feedback :)

Commit:
Add Gradle support

Commit:
Implement level-1 of individual project

Added the following functionalities to the main() function:
* Logo & introduction to app
* Echo/print out user input
* Exit app when user types "bye"

Commit:
Implement level 2 of individual project

New features:
*App now keeps track of items entered by user, up to a number of 100 items
*App can list items entered by user if they type 'list'

Edited:
*Starting instructions to describe new features

Commit:
Implement level-3 of individual project

- Added in Task class, to help keep track of items as tasks and mark them as 'done'
- Added in new instructions related to marking tasks as 'done'

Commit:
Implement level 4 of individual project

Added in different task subclasses, making the task list more customisable and versatile
-ToDo tasks are similar to tasks in the previous level, just a description
-Deadline tasks will have an additional deadline time parameter, represented as a String for now
-Event tasks have a time range parameter, also represented as a String for now

Added in a help command to guide the user on valid commands.

Commit:
Implement A-TextUiTesting of individual project

Added in some sample inputs and outputs to input.txt and EXPECTED.txt respectively to semi-automate the testing of the Duke Bot.

Commit:
Implemented level-5 of individual project

Added various error handling to ensure the bot runs smoothly without breaking, these include checking for:
-Invalid/Out-of-bounds integers when marking tasks as done
-Unrecognised commands when the bot is idle
-Not having a description for tasks

Commit:
Implement level-6 of individual project

Added the feature to delete tasks from the task list.
This would be useful when the user completes a task and/or wants to make a new list of tasks
-The task list has been changed to an ArrayList type instead of a basic array.
-The limit of 100 tasks is still maintained to prevent the list from being too long.

botr99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,335 @@
+
 public class Duke {
+    //Task and Subclasses
+    public static class Task{
```
Perhaps add a space before `{`? Noticed this in other places as well

Zacchua replied:
Perhaps these nested classes should be extracted? Current Duke class seems abit bloated

botr99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,335 @@
+        //write all tasks to the file. Separated by new lines
+        try {
+            FileWriter myWriter = new FileWriter("./duke.txt");
+            for (int i = 1; i < tasks.size()+1; i++) {
+                myWriter.write(i + ". " + tasks.get(i-1).toString());
```
Perhaps add a space between the operators? Noticed this in other places as well

botr99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,335 @@
+                                System.out.println("Enter date/time in this exact format: dd/mm/yyyy hhmm");
+                                input = reader.nextLine();
+                                try{
+                                    int day = Integer.parseInt(input.substring(0,2));
+                                    int month = Integer.parseInt(input.substring(3,5));
+                                    int year = Integer.parseInt(input.substring(6,10));
+
+                                    int hours = Integer.parseInt(input.substring(11,13));
+                                    int minutes = Integer.parseInt(input.substring(13));
+                                    date = java.time.LocalDate.of(year,month,day);
+                                    time = java.time.LocalTime.of(hours,minutes);
```
Perhaps consider using `DateTimeFormatter` to parse the date and time?

botr99 reviewed (commented):
Overall, except for white space within statements, most of the coding standards are adhered to

Zacchua commented on duke.txt:
```diff
@@ -0,0 +1,2 @@
+1. [T] [ ] eat lunch
```
This seems like the file where you store the tasks. Should this be included in the commit?

Zacchua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,335 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        //START OF PROGRAM
```
This main function seems quite bloated as well, perhaps it should be broken up into separate functions?

Zacchua reviewed (commented):
Interesting approach, I also like that you have a summary for the all the commands and how to use them, as well as a help command

Commit:
Implement level-7 of individual project

Added ability for bot to save task list to a file every time the task list is updated
-The file is hardcoded to "./duke.txt"

Commit:
Implement level-8 of individual project

Added ability for bot to recognise dates and times
-Program now properly formats string inputs for dates and times
-Deadline and Event tasks now take in LocalDate and LocalTime parameters

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Commit:
Implement A-OOP for individual project

Enhanced the OOP of the bot, improving code maintainability and quality
-Code is now modelled into 4 classes: Ui, Parser, TaskList and Storage

Added ability to load tasks from Duke.txt which was missing in Level-7

Commit:
Implement A-Packages of individual project

-Put all classes into one complete package "duke"

-Minor bug fixes for Ui prompts

Commit:
Implement A-JUnit of individual project

-Created a JUnit class DukeTest with several tests to test various methods
These include Task creation, Storage.load() and TaskList.getTask()

-Updated .gitignore to ignore some extra files

Commit:
Implement A-Jar of individual project

-Packaged the project into a JAR file "ip.jar" for convenient distribution and use

-Updated .gitignore

Commit:
Implement A-JavaDoc for individual project

Added JavaDoc documentation to elaborate on Duke classes and methods

Updated .gitignore to ignore Duke.txt

Commit:
Implement A-CodingStandard of individual project

Improved code readability by following Java Coding Standard
-Fixed indentation of switch case blocks
-Added spaces between parameters and commas for functions
-Fixed text wrapping issues

Commit:
Implement level-9 of individual project

Added the ability for the bot to search for specific tasks based on keywords in their descriptions

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java

EssWhyy committed:
Delete duke.txt

Commit:
Improve code readability and quality

-Based on given Java Code Standard

Commit:
Merge branch 'master' of github.com:EssWhyy/ip

Commit:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

Commit:
Divide classes into packages

Grouped classes into separate packages to improve code readability and maintainability

Commit:
Implement level-10 of individual project

Converted the text UI to a graphical UI to make the bot more user friendly and appealing
-Added JavaFX libraries
-Created java and fxml files for Dialogs and Window
-Refactored some code in Parser.java

Deleted Text Ui (Ui.java) and files from A-TextUiTesting as they are no longer necessary

Commit:
Merge branch 'branch-Level-10'

Commit:
Implement A-CodeQuality of individual project

Currently the code for the bot across multiple classes is not entirely refined and is a bit messy.

The code is improved based on the code quality topics learnt in CS2103T so as to make it more readable and maintainable.

This includes separating parts of long methods out, adding more javadoc comments, and minor bug fixes and optimisations.

Commit:
Implement A-Assertions of individual project

Used assert to confirm if TaskList is not empty

EssWhyy committed:
Merge pull request #3 from EssWhyy/branch-A-Assertions

Implement A-Assertions of individual project

EssWhyy committed:
Merge pull request #2 from EssWhyy/branch-A-CodeQuality

Implement A-CodeQuality of individual project

Commit:
Implement C-Sort extension for individual project

Right now, there is no way to manipulate the task list
besides adding and deleting a task.

This can make the task list messy and difficult to track tasks
after multiple insertions or deletions.

A sort function is implemented which allows the user
to sort the task list by task and date/time.

This allows them to separate out different types of tasks,
and see which deadlines/events are coming up soon.

Commit:
Merge branch 'branch-C-Sort'

Commit:
Prevent addition of more tasks when task list size is more than 100

Commit:
Add Ui.png

Minor enhancements to java and fxml code

Commit:
Add Ui.png

Commit:
Implement A-UserGuide of individual project

Created a user guide to help users understand and use the bot.
